Shell/Shell.cpp

```diff
@@ -981,20 +1046,33 @@ Vector<String> Shell::process_arguments(const Vector<Token>& args)
         if (arg.type == Token::Comment)
             continue;
 
+        if (arg.type == Token::SingleQuoted) {
+            argv_string.append(arg.text);
+            continue;
+        }
+
         // This will return the text passed in if it wasn't a variable
         // This lets us just loop over its values
-        auto expanded_parameters = expand_parameters(arg.text);
+        bool is_quoted = arg.type == Token::DoubleQuoted;
+        auto expanded_parameters = expand_parameters(arg.text, is_quoted);
 
+        // Bash-ism:
+        // Expanded variables will be expanded again if they contain globs and are not quoted.
```

Comment:
In my mind, by the time we get to applying globs, the variables should be already gone, and we shouldn't have to care about them at that stage. This comment seems to indicate we do have to care?

Reply:
That's the thing, _because_ we don't care about them, things like `$foo$bar` expanding to form a glob (and cause a fair bit of confusion - ask stackoverflow) are created.
ideally, the result of a variable substitution wouldn't be expanded _again_, at least, in my ideal world.

Reply:
Clearly, `$var` substitution and glob expansion are separate things done at different stages, at least in Bash.

In my ideal world — not in a shell language, perhaps — I would be explicit in specifying when I want something substituted and/or something globbed. But that requires quoting "plain text", which is undesirable for a shell. E.g. I'd want `"foo", bar, glob("*.txt"), glob(meh)` instead of `foo "$bar" *.txt $meh` — explicit, unambiguous, and too clumsy for a shell.

Reply:
Enter zsh!
if you give it a glob, you clearly want it expanded.
if you quote the glob, you clearly don't.
if you substitute in a variable, you probably don't want it glob-expanded, if you do, explicitly tell the shell that you want further expansion after the substitution: `$~var`

But I doubt we're trying to go there with this shell, so I think sticking to the (bad) semantics of bash is the way to go, unless they're _really_ bad

Reply:
but what if you have a [single/double quoted?] variable that expands to a [single/double quoted?] glob that expands to a [single/double quoted?] variable... This gets insane quick

Reply:
Single quoted strings are never expanded, they're verbatim.

I'm not sure how zsh handles the insanity that you're brewing, but I imagine it would apply one more round of expansions because of "$~" and then just stop.

I'm pretty sure (at least in zsh) variables cannot expand to form other variables (would be cool, `export quine="export quine=\"$quine\""`)

Reply:
Perhaps we should decide what are our goals here. Do we want to build a "sane" shell? Do we want to mimic Bash? Do we want to implement what POSIX specifies for a shell? Do we want to eventually run scripts written for Bash and/or POSIX shell using our Shell?

Depending on the answers, we would implement the semantics / behavior that we want. I, for one, do like the idea of extra splitting / substitution / globs never being implicitly performed on results of other expansions/globs, if we go our own way instead of mimicking Bash.

Reply:
I personally would prefer a sane shell, and I've expressed that I dislike how bash does some things before...
I, for one, wouldn't mind throwing bash-compat out the window, we already have a bash port for that.

but for goals...let's also ask our good friend @awesomekling about them.

Reply:
Goal: A powerful shell that feels nice to use and doesn't do stupid shit you don't want :)

Reply:
So I guess we're doing our own thing then.

so far as this PR goes, I think I'll just make it not expand more than once, and we can figure out the correct way once someone gives some love to the parser and gets it to spit out good ASTs.

Reply:
So if we're talking a sane, nice shell... There are:
* Byte strings (can have any bytes inside, including whitespace)
* Lists of byte strings

You can go from a string to a list of strings (splitting them). You can concatenate strings and lists. You can expand a variable (to a string or to a list of strings? — to be determined). You can glob; the glob itself is always a string, its result is always a list.

Then we want syntax for all this; and we want that syntax to be the same as good old shell in simple cases.

Right?

Reply:
Hmm...let's go through those,

- byte strings: I don't really mind, but I don't see how useful they'd really be over just `\xab`
- lists of X: much prefer a way to have good types, and somehow maintain the "can treat anything as a string" feel in normal cases
   - not as _in-depth_ as other typed shells (powershell comes to mind)
   - not as stringly-typed as some other shells either.
- string to list of strings: I really like this, it's a very useful feature, would simplify looping semantics as well.
- concat string list: uhh...I guess? I'd rather have better list operations, I wouldn't mind saying `$some_element:$some_list`.
- well-defined globs without bullshit: absolutely!

Syntax for this stuff...indeed, that's a whole design session waiting to happen.

Reply:
> byte strings: I don't really mind, but I don't see how useful they'd really be over just `\xab`

No no, I'm not suggesting a syntax for string literals; `\xab` would probably be the syntax indeed. I'm saying that internally the Shell should clearly keep track of what is just a byte string — which can have bytes inside, including invalid UTF-8 and whitespace, and those spaces would never implicitly cause the string to be expanded to a list — and what is a list.

> lists of X: much prefer a way to have good types, and somehow maintain the "can treat anything as a string" feel in normal cases

Yes, this is what I'm suggesting — provide clear semantics for what is a string and what is a list and how to go between the two, and make it look & feel "as usual shell" in simple cases.

> concat string list: uhh...I guess?

In say Bash concatting two lists is done just by writing them next to each other in a list context, e.g. `echo $foo $bar`. Again, I'm not saying we should use that syntax, or that we should not; I'm saying we should acknowledge that this is an operation we want to support.

Reply:
Ah, I see.
the shell doesn't really care what encoding its prompt has, it's just that libline thinks "must be valid unicode" about its lines.
I'm not sure if we can have the best of both worlds like that, we can either not care about the encodings, and lose unicode, or care about the encodings and lose raw bytestring inputs.
Of course, if the syntax is strictly \xab, we can still live in the blissful unicode land.

Other points, we agree on.

Reply:
Ah, I see.
the shell doesn't really care what encoding its prompt has, it's just that libline thinks "must be valid unicode" about its lines.
I'm not sure if we can have the best of both worlds like that, we can either not care about the encodings, and lose unicode, or care about the encodings and lose raw bytestring inputs.
Of course, if the syntax is strictly \xab, we can still live in the blissful unicode land.

Other points, we agree on.

Reply:
what the heck, duplication of messages, in the year of our lord 2020?

Reply:
> the shell doesn't really care what encoding its prompt has, it's just that libline thinks "must be valid unicode" about its lines.

Those invalid UTF-8 things don't have to be typed in in the interactive shell (in fact it likely wouldn't be, because how'd you do that), it may appear internally after some expansions / globs.

Reply:
do we really want to expand and let the user still edit the result?
whatever happens as the result of doing things with the user input should remain inside the shell, I think.

Reply:
No, that's what I'm saying, the result of expansion only lives inside the shell, and never makes it to the TTY.

Reply:
fantastic, we're saying the same thing lol.